The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27341219796/job/80778358725) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/80778358725) [(plain)](https://github.com/rust-lang/rust/commit/c65f522676f475e47b248c1cafed0c49f30fb161/checks/80778358725/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6900 files
tidy check
tidy [deps]: proc-macro crate dependency `jobserver` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc-stable-hash` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_aarch64_msvc` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc_index` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc_data_structures` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `elsa` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `pin-project-lite` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `r-efi` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `wasi` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `redox_syscall` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `psm` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `memmap2` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_x86_64_gnullvm` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `shlex` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `errno` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_i686_gnu` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-link` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rand` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rand_core` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rand_xorshift` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-targets` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc_arena` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-strings` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `either` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `stacker` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `scoped-tls` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-sys` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-future` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `wit-bindgen-rt` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-collections` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `parking_lot` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `arrayvec` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rand_chacha` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_x86_64_msvc` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_i686_gnullvm` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `ena` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `fastrand` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc_thread_pool` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `thin-vec` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-result` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `parking_lot_core` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-numerics` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `crossbeam-utils` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `zerocopy` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `measureme` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `linux-raw-sys` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc_graphviz` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `cc` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `tracing` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `once_cell` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_i686_msvc` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `tracing-core` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustix` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `perf-event-open-sys` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `scopeguard` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `ppv-lite86` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `portable-atomic` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `getrandom` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_x86_64_gnu` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `tempfile` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-threading` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc_serialize` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `valuable` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `crossbeam-deque` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `crossbeam-epoch` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows_aarch64_gnullvm` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `lock_api` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `windows-core` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: proc-macro crate dependency `rustc_hashes` is not registered in `src/bootstrap/src/utils/proc_macro_deps.rs`
tidy [deps]: Run `./x.py test tidy --bless` to regenerate the list
tidy [deps]: FAIL
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
---
linting javascript files
Running eslint on rustdoc JS files
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: deps
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1624:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1511:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1511:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1596:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1433:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1139:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1118:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:02:50
  local time: Thu Jun 11 10:47:41 UTC 2026
  network time: Thu, 11 Jun 2026 10:47:41 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/compiler/rustc_type_ir_macros/Cargo.toml
+++ b/compiler/rustc_type_ir_macros/Cargo.toml
@@ -11,6 +11,7 @@ nightly = []
 
 [dependencies]
 # tidy-alphabetical-start
+indexmap = "2.4.0"
 proc-macro2 = "1"
 quote = "1"
 syn = { version = "2.0.9", features = ["full", "visit-mut"] }

--- a/compiler/rustc_type_ir_macros/src/lib.rs
+++ b/compiler/rustc_type_ir_macros/src/lib.rs
@@ -1,3 +1,4 @@
+use indexmap::IndexSet;
 use quote::{ToTokens, quote};
 use syn::visit_mut::VisitMut;
 use syn::{Attribute, parse_quote};
@@ -17,10 +18,23 @@ decl_derive!(
     [GenericTypeVisitable] => customizable_type_visitable_derive
 );
 
-struct LiftedTy {
+struct TransformedTy {
     ty: syn::Type,
-    generic_parameter_bounds: Vec<syn::Ident>,
+    generic_parameter_bounds: IndexSet<syn::Ident>,
 }
+
+enum TypeParameterPath {
+    Interner,
+    GenericParameter(syn::Ident),
+}
+
+enum TypeParameterTransform {
+    Continue,
+    Stop,
+}
+
+type TypeParameterVisitor =
+    fn(TypeParameterPath, &mut syn::TypePath, &mut IndexSet<syn::Ident>) -> TypeParameterTransform;
 
 fn has_ignore_attr(attrs: &[Attribute], name: &'static str, meta: &'static str) -> bool {
     let mut ignored = false;
@@ -91,6 +105,9 @@ fn type_foldable_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::Toke
 
     s.add_where_predicate(parse_quote! { I: Interner });
     s.add_bounds(synstructure::AddBounds::Fields);
+    let generic_parameters =
+        s.ast().generics.type_params().map(|ty| ty.ident.clone()).collect::<Vec<_>>();
+    let mut generic_parameter_bounds = IndexSet::new();
     s.bind_with(|_| synstructure::BindStyle::Move);
     let body_try_fold = s.each_variant(|vi| {
         let bindings = vi.bindings();
@@ -101,6 +118,12 @@ fn type_foldable_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::Toke
             if has_ignore_attr(&bind.ast().attrs, "type_foldable", "identity") {
                 bind.to_token_stream()
             } else {
+                for param in
+                    type_foldable_generic_parameters(bind.ast().ty.clone(), &generic_parameters)
+                {
+                    generic_parameter_bounds.insert(param);
+                }
+
                 quote! {
                     ::rustc_type_ir::TypeFoldable::try_fold_with(#bind, __folder)?
                 }
@@ -129,6 +152,9 @@ fn type_foldable_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::Toke
     // to generate code for them.
     s.filter(|bi| !has_ignore_attr(&bi.ast().attrs, "type_foldable", "identity"));
     s.add_bounds(synstructure::AddBounds::Fields);
+    for param in generic_parameter_bounds {
+        s.add_where_predicate(parse_quote! { #param: ::rustc_type_ir::TypeFoldable<I> });
+    }
     s.bound_impl(
         quote!(::rustc_type_ir::TypeFoldable<I>),
         quote! {
@@ -147,6 +173,19 @@ fn type_foldable_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::Toke
             }
         },
     )
+}
+
+fn type_foldable_generic_parameters(
+    ty: syn::Type,
+    generic_parameters: &[syn::Ident],
+) -> IndexSet<syn::Ident> {
+    transform_type_parameters(ty, generic_parameters, |path, _, generic_parameter_bounds| {
+        if let TypeParameterPath::GenericParameter(param) = path {
+            generic_parameter_bounds.insert(param);
+        }
+        TypeParameterTransform::Continue
+    })
+    .generic_parameter_bounds
 }
 
 /// `Lift_Generic` is specialised for structs/enums parameterised by an interner
@@ -251,40 +290,72 @@ fn is_type_phantom(ty: &syn::Type) -> bool {
     get_first_path_segment(ty).is_some_and(|segment| segment.ident == "PhantomData")
 }
 
-fn lift(mut ty: syn::Type, generic_parameters: &[syn::Ident]) -> LiftedTy {
-    struct ItoJ<'a> {
+fn lift(ty: syn::Type, generic_parameters: &[syn::Ident]) -> TransformedTy {
+    transform_type_parameters(ty, generic_parameters, |path, ty, generic_parameter_bounds| {
+        match path {
+            TypeParameterPath::Interner => {
+                *ty.path.segments.first_mut().unwrap() = parse_quote! { J };
+                TypeParameterTransform::Continue
+            }
+            TypeParameterPath::GenericParameter(param) => {
+                generic_parameter_bounds.insert(param.clone());
+                *ty = parse_quote! { <#param as ::rustc_type_ir::lift::Lift<J>>::Lifted };
+                TypeParameterTransform::Stop
+            }
+        }
+    })
+}
+
+fn transform_type_parameters(
+    mut ty: syn::Type,
+    generic_parameters: &[syn::Ident],
+    visit: TypeParameterVisitor,
+) -> TransformedTy {
+    struct TypeParameterTransformer<'a> {
         generic_parameters: &'a [syn::Ident],
-        generic_parameter_bounds: Vec<syn::Ident>,
+        generic_parameter_bounds: IndexSet<syn::Ident>,
+        visit: TypeParameterVisitor,
     }
 
-    impl VisitMut for ItoJ<'_> {
+    impl VisitMut for TypeParameterTransformer<'_> {
         fn visit_type_path_mut(&mut self, i: &mut syn::TypePath) {
-            if i.qself.is_none() {
+            let path = if i.qself.is_none() {
                 let segments_len = i.path.segments.len();
-                if let Some(first) = i.path.segments.first_mut() {
-                    // Turn paths from `I` into `J`
+                i.path.segments.first().and_then(|first| {
                     if first.ident == "I" {
-                        *first = parse_quote! { J };
+                        Some(TypeParameterPath::Interner)
                     } else if segments_len == 1
                         && matches!(first.arguments, syn::PathArguments::None)
-                        && self.generic_parameters.iter().any(|param| first.ident == *param)
+                        && self.generic_parameters.contains(&first.ident)
                     {
-                        let ident = first.ident.clone();
-                        if !self.generic_parameter_bounds.iter().any(|param| *param == ident) {
-                            self.generic_parameter_bounds.push(ident.clone());
-                        }
-
-                        *i = parse_quote! { <#ident as ::rustc_type_ir::lift::Lift<J>>::Lifted };
-                        return;
+                        Some(TypeParameterPath::GenericParameter(first.ident.clone()))
+                    } else {
+                        None
                     }
+                })
+            } else {
+                None
+            };
+
+            if let Some(path) = path {
+                if let TypeParameterTransform::Stop =
+                    (self.visit)(path, i, &mut self.generic_parameter_bounds)
+                {
+                    return;
                 }
             }
+
             syn::visit_mut::visit_type_path_mut(self, i);
         }
     }
-    let mut visitor = ItoJ { generic_parameters, generic_parameter_bounds: Vec::new() };
+
+    let mut visitor = TypeParameterTransformer {
+        generic_parameters,
+        generic_parameter_bounds: IndexSet::new(),
+        visit,
+    };
     visitor.visit_type_mut(&mut ty);
-    LiftedTy { ty, generic_parameter_bounds: visitor.generic_parameter_bounds }
+    TransformedTy { ty, generic_parameter_bounds: visitor.generic_parameter_bounds }
 }
 
 #[cfg(not(feature = "nightly"))]

